proposals/0000-no-fallible-do.rst

```diff
@@ -0,0 +1,261 @@
+It is far from clear whether we want the unfolding of syntax sugar to depend on anything that is going on at the type level in the first place, as that makes it far more challenging to teach and understand.
+Monads are already an infamous steep part of the learning curve without having to drag in the entire architecture of the compiler frontend to explain ``do``\ -notation.
+
+Avoid ``fail`` altogether
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Looking at the problem as described above, our options are limited.
+``fail`` worked well before because its presence or absence did not influence type checking.
+At the same time, we do not want to go back to putting ``fail`` back inside ``Monad``.
+The simplest thing to do is just cut the Gordian knot, and provide a way to not use ``fail`` in the desugaring.
+
+This may sound drastic, even taking into account we'll propose something opt-in rather than a breaking change in the next section.
+But, we'd like the case that given the way Haskell is actually written, it isn't.
+Often, one has no interest in having ``MonadFail`` constraints appear at all.
+Errors made with ``fail`` are always strings, and so completely unstructured.
+Whether one prefers ``Either`` and ``EitherT`` or synchronous exceptions, the norm is to use types to structure failure modes, just as we use types to structure everything else in Haskell.
+Furthermore, even if one does want to use unstructured textual errors, ``fail`` uses ``String`` rather than something with better performance characteristics like ``Text``.
+For these reasons, one might want to avoid going down a rabbit hole of subtler issues of completeness checking when the real immediate problem at hand is ``fail`` ever being used in the translation at all.
+
+Proposed Change Specification
+-----------------------------
+
+We propose a module-level means of switching off the use of ``fail`` in ``do``\ -syntax altogether via an extension flag.
+Specifically, there is a extension flag ``FallibleDo``, which is enabled by default, that controls the desugaring of fallible patterns in ``do``\ -notation binds (i.e. ``pat <- stmt`` syntax).
+
+With ``FallibleDo``, the usual translation of the ``do``\ -syntax involving ``fail`` is used.
+With ``NoFallibleDo`` the use of ``fail`` is replaced with the usual throwing of a `PatternMatchFail <https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Exception.html#t:PatternMatchFail>`_.
+
+When the ``-Wincomplete-uni-patterns`` warning flag is enabled alongside ``NoFallibleDo``, we will warn about the incomplete pattern match.
+
+Monad comprehensions are not affected by this extension.
+
+Examples
+--------
+
+If we take the example from the Motivation section:
+
+::
+
+  example x = do
+    A n <- x
+    return n
+
+With FallibleDo turned on (the default), this presently translates to:
+
+::
+
+  example x = x >>= \v -> case v of
+    A n -> return n
+    _ -> fail "..."
+
+which has an inferred type which is constrained by ``MonadFail m``
```

Comment:
Oh this was just fast and loose what it "morally" desugars too / what it "means". Not trying to propose how things actually work in GHC at this point of the motivation.

Happy to reword to make it clearer.

Reply:
If the example is not the "specification" of the change, then it would help if you say more than simply

> With FallibleDo, the usual translation of the do-syntax involving fail is used. With NoFallibleDo the use of fail is replaced with the usual throwing of a PatternMatchFail.

I think a good way would be a diff in terms of https://www.haskell.org/onlinereport/haskell2010/haskellch3.html#x8-470003.14: Specifically, instead of

```
do {p <- e; stmts} =
  let ok p = do {stmts}
      ok _ = fail "..."
  in e >>= ok
```

you want

```
do {p <- e; stmts} =
  let ok p = do {stmts}
  in e >>= ok
```

with an implicit pattern match failure in the second translation.